src/renderer/base/FontResource.cpp

```diff
@@ -0,0 +1,285 @@
+    static constexpr DWORD DFF_FIXED = 0x0001;
+    static constexpr DWORD DFF_PROPORTIONAL = 0x0002;
+    static constexpr DWORD DFF_1COLOR = 0x0010;
+    static constexpr DWORD DFF_16COLOR = 0x0020;
+    static constexpr DWORD DFF_256COLOR = 0x0040;
+    static constexpr DWORD DFF_RGBCOLOR = 0x0080;
+
+    // DRCS soft fonts only require 96 characters at most.
+    static constexpr size_t CHAR_COUNT = 96;
+
+#pragma pack(push, 1)
+    struct GLYPHENTRY
+    {
+        WORD geWidth;
+        DWORD geOffset;
+    };
+
+    struct FONTINFO
+    {
+        WORD dfVersion;
+        DWORD dfSize;
+        CHAR dfCopyright[60];
+        WORD dfType;
+        WORD dfPoints;
+        WORD dfVertRes;
+        WORD dfHorizRes;
+        WORD dfAscent;
+        WORD dfInternalLeading;
+        WORD dfExternalLeading;
+        BYTE dfItalic;
+        BYTE dfUnderline;
+        BYTE dfStrikeOut;
+        WORD dfWeight;
+        BYTE dfCharSet;
+        WORD dfPixWidth;
+        WORD dfPixHeight;
+        BYTE dfPitchAndFamily;
+        WORD dfAvgWidth;
+        WORD dfMaxWidth;
+        BYTE dfFirstChar;
+        BYTE dfLastChar;
+        BYTE dfDefaultChar;
+        BYTE dfBreakChar;
+        WORD dfWidthBytes;
+        DWORD dfDevice;
+        DWORD dfFace;
+        DWORD dfBitsPointer;
+        DWORD dfBitsOffset;
+        BYTE dfReserved;
+        DWORD dfFlags;
+        WORD dfAspace;
+        WORD dfBspace;
+        WORD dfCspace;
+        DWORD dfColorPointer;
+        DWORD dfReserved1[4];
+        GLYPHENTRY dfCharTable[CHAR_COUNT];
+        CHAR szFaceName[LF_FACESIZE];
+        BYTE bitmaps[1];
+    };
+#pragma pack(pop)
+}
+
+FontResource::FontResource(const gsl::span<const uint16_t> bitPattern,
+                           const til::size sourceSize,
+                           const til::size targetSize,
+                           const size_t centeringHint) :
+    _bitPattern{ bitPattern.begin(), bitPattern.end() },
+    _sourceSize{ sourceSize },
+    _targetSize{ targetSize },
+    _centeringHint{ centeringHint }
+{
+}
+
+void FontResource::SetTargetSize(const til::size targetSize)
+{
+    if (_targetSize != targetSize)
+    {
+        _targetSize = targetSize;
+        _fontHandle = nullptr;
+    }
+}
+
+FontResource::operator HFONT()
+{
+    if (!_fontHandle && !_bitPattern.empty())
+    {
+        _regenerateFont();
+    }
+    return _fontHandle.get();
+}
+
+void FontResource::_regenerateFont()
+{
+    const auto targetWidth = _targetSize.width<WORD>();
+    const auto targetHeight = _targetSize.height<WORD>();
+    const auto charSizeInBytes = (targetWidth + 7) / 8 * targetHeight;
+
+    const DWORD fontBitmapSize = charSizeInBytes * CHAR_COUNT;
+    const DWORD fontResourceSize = sizeof(FONTINFO) + fontBitmapSize;
+
+    auto fontResourceBuffer = std::vector<byte>(fontResourceSize);
+    void* fontResourceBufferPointer = fontResourceBuffer.data();
+    auto& fontResource = *static_cast<FONTINFO*>(fontResourceBufferPointer);

```

Comment:
Yeah, this makes sense to me. I can't use `unique_ptr` and `reinterpret_cast` because those will both fail an audit build, but I've now got rid of the 1-element array, and just set `dfBitsOffset` to `sizeof(FONTINFO)`. I realised I could then easily create an appropriate span for the `_resizeBitPattern` call by getting a subspan of the `fontResourceBuffer`, and that sets the size automatically.